### PR TITLE
Query AWS to find subnets with explicitely specified subnet IDs

### PR DESCRIPTION
**Pull request by codablock**
**What type of PR is this?**
/kind feature

**What this PR does / why we need it**:
This basically unifies the code used to find subnets with filters and subnets
by ID. This is done by handling the ID case as a simple filter instead of
looking into the cluster network spec.

The reason is that explicitely specified subnets are not necessarely part
of the cluster network spec. For example, it might be desired to run a
EKS control plane inside 3 specific subnets (so they are part of the
network spec) while running workers in different subnets from the same
VPC.

A side effect of this change is that we don't have to check for the correct
failure domain and/or public/private subnets anymore as this is implicitely
handled by the criterias.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Checklist**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests
- [ ] adds or updates e2e tests

**Release note**:

```release-note
Allow to specify subnets by ID which are not part of the cluster's network spec
```

### CODE REVIEW

**Comment by enxebre** [comment]
This make sense to me.
Another example where this is natural is externally managed infrastructure. I see this decoupling as a good thing.

**Comment by codablock** [comment]
@randomvariable Sorry to bother you, but did you find time to look into this? :)

**Comment by enxebre** [comment]
cc @richardcase @sedefsavas for awareness, any objections to proceed with this?

**Comment by richardcase** [comment]
> cc @richardcase @sedefsavas for awareness, any objections to proceed with this?

Will take a look @enxebre

**Review by sedefsavas — COMMENTED** [review]
Thanks for cleaning up this logic @codablock.
This is actually a bug instead of a feature.

We fixed a very similar issue with ELBs in this PR https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2877
/assign @pydctw for lgtm.

File: pkg/cloud/services/ec2/instances.go [hunk 1]
```diff
@@ -260,57 +260,56 @@ func (s *Service) findSubnet(scope *scope.MachineScope) (string, error) {
...
+				scope.Name(), criteria)
+			record.Warnf(scope.AWSMachine, "FailedCreate", errMessage)
+			return "", awserrors.NewFailedDependency(errMessage)
+		}
+
+		var filtered []*ec2.Subnet
+		errMessage := fmt.Sprintf("failed to run machine %q, found %d subnets matching criteria but post-filtering failed.",
+			scope.Name(), len(subnets))
+		for _, subnet := range subnets {
+			if failureDomain != nil && *subnet.AvailabilityZone != *failureDomain {
```
[the remark this answers is not in the document]
Comment (codablock) [reply pkg/cloud/services/ec2/instances.go:297]: This is what I did in the initial solution, but then realised that the error message returned by EC2 would be cryptic (subnet not found) and not really say what is really wrong (subnet found, but mismatching failure domain).

**Comment by Ankitasw** [comment]
> Thanks for cleaning up this logic @codablock.
> This is actually a bug instead of a feature.
> 
> We fixed a very similar issue with ELBs in this PR https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/2877
> /assign @pydctw for lgtm.

@pydctw gentle reminder for review.

**Comment by pydctw** [comment]
Other than a minor comment on the test, look good to me. Thanks for simplifying findSubnet logic by combining two cases.

Could you also squash the commits?

File: pkg/cloud/services/ec2/instances_test.go [hunk 2]
```diff
@@ -1439,9 +1598,24 @@ func TestCreateInstance(t *testing.T) {
...
+					DescribeSubnets(&ec2.DescribeSubnetsInput{
+						Filters: []*ec2.Filter{
+							filter.EC2.SubnetStates(ec2.SubnetStatePending, ec2.SubnetStateAvailable),
+							filter.EC2.VPC("vpc-id"),
+							{Name: aws.String("subnet-id"), Values: aws.StringSlice([]string{"private-subnet-1"})},
+						},
+					}).
+					Return(&ec2.DescribeSubnetsOutput{
+						Subnets: []*ec2.Subnet{{
+							SubnetId:            aws.String("public-subnet-1"),
```
[the remark this answers is not in the document]
Comment (codablock) [reply pkg/cloud/services/ec2/instances_test.go]: True, fixed it :+1:

**Comment by codablock** [comment]
@pydctw @sedefsavas Thanks for you reviews :)
I've rebased on main, fixed the last nit and squashed commits. Hopefully ready to merge now.

**Pull request merged**